src/pyk/kcfg/kcfg.py

```diff
@@ -968,6 +968,31 @@ def prune(self, node_id: str) -> None:
         for node in nodes:
             self.remove_node(node.id)
 
+    def shortest_path_between(self, source_node_id: str, target_node_id: str) -> Iterable[Successor] | None:
+        paths = self.paths_between(source_node_id, target_node_id)
+        return sorted(paths, key=(lambda path: len(path)))[0]
+
+    def path_constraints(self, final_node_id: str) -> list[KInner]:
+        cover_substs: list[CSubst] = []
+        constraints: list[KInner] = []
+
+        def apply_substs(constraint: KInner) -> KInner:
+            curr_constraint = constraint
+            for subst in cover_substs:
+                curr_constraint = subst.apply(CTerm.from_kast(curr_constraint)).kast
+            return constraint
+
+        path = self.shortest_path_between(self.get_unique_init().id, final_node_id)
+        if path is None:
+            raise ValueError(f'No path found to specified node: {final_node_id}')
+        for edge in path:
+            if type(edge) is KCFG.Split:
+                for _, csubst in edge.targets:
+                    constraints.append(apply_substs(csubst.constraint))
+            if type(edge) is KCFG.Cover:
+                cover_substs.append(edge.csubst)
+        return constraints
```

Comment:
```suggestion
        curr_constraint = mlTop()
        curr_subst = Subst({})
        for edge in reversed(path):
            if type(edge) is KCFG.Split:
                assert len(edge.targets) == 1
                _, csubst = edge.targets[0]
                curr_constraint = mlAnd([curr_subst(csubst.constraint), curr_constraint])
                curr_subst = csubst.subst(curr_subst)
            if type(edge) is KCFG.Cover:
                curr_constraint = edge.csubst(curr_constraint)
        return curr_constraint
```

Perhaps this is simpler, and doesn't require the weird stateful `apply_substs` method?

Reply:
It's simpler and looking at this now I realize my version needs to be applying the substitutions in  the other order. I'm not sure what the line
```
                curr_subst = csubst.subst(curr_subst)
```
is doing, though. The way I understand, the substitution gets applied when we hit it going backwards, and so there would be no need to chain it onto the current substitution. The version I've pushed just now is:
```
        curr_constraint: KInner = mlTop()
        for edge in reversed(path):
            if type(edge) is KCFG.Split:
                assert len(edge.targets) == 1
                _, csubst = edge.targets[0]
                curr_constraint = mlAnd([csubst.constraint, curr_constraint])
            if type(edge) is KCFG.Cover:
                curr_constraint = edge.csubst.apply(CTerm.from_kast(curr_constraint)).kast
        return curr_constraint